modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/session/Session.java

```diff
@@ -0,0 +1,168 @@
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.apache.ignite.internal.sql.engine.AsyncCloseable;
+import org.apache.ignite.internal.sql.engine.CurrentTimeProvider;
+import org.apache.ignite.internal.sql.engine.property.PropertiesHolder;
+
+/**
+ * A session object.
+ *
+ * <p>This is a server-side sql session which keeps track of associated resources like opened query cursor, or keeps a properties holder
+ * with all properties set during session's creation.
+ */
+public class Session implements AsyncCloseable {
+    /** Marker used to mark a session which has been experid. */
+    private static final long EXPIRED = 0L;
+
+    private final Set<AsyncCloseable> resources = Collections.synchronizedSet(
+            Collections.newSetFromMap(new IdentityHashMap<>())
+    );
+
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
+
+    private final AtomicReference<CompletableFuture<Void>> closeFutRef = new AtomicReference<>();
+
+    private final long idleTimeoutMs;
+    private final SessionId sessionId;
+    private final AtomicLong lastTouched;
+    private final PropertiesHolder queryProperties;
+    private final CurrentTimeProvider currentTimeProvider;
+
+    /**
+     * Constructor.
+     *
+     * @param sessionId A session identifier.
+     * @param currentTimeProvider The time provider used to update the timestamp on every touch of this object.
+     * @param idleTimeoutMs Duration in milliseconds after which the session will be considered expired if no action have been
+     *                      performed on behalf of this session during this period.
+     * @param queryProperties The properties to keep within.
+     */
+    public Session(
+            SessionId sessionId,
+            CurrentTimeProvider currentTimeProvider,
+            long idleTimeoutMs,
+            PropertiesHolder queryProperties
+    ) {
+        this.sessionId = sessionId;
+        this.currentTimeProvider = currentTimeProvider;
+        this.idleTimeoutMs = idleTimeoutMs;
+        this.queryProperties = queryProperties;
+
+        lastTouched = new AtomicLong(currentTimeProvider.now());
+    }
+
+    /** Returns the properties this session associated with. */
+    public PropertiesHolder queryProperties() {
+        return queryProperties;
+    }
+
+    /** Returns the identifier of this session. */
+    public SessionId sessionId() {
+        return sessionId;
+    }
+
+    /** Checks whether the given session has expired or not. */
+    public boolean expired() {
+        var last = lastTouched.get();
+
+        if (last == EXPIRED) {
+            return true;
+        }
+
+        return currentTimeProvider.now() - last > idleTimeoutMs && lastTouched.compareAndSet(last, EXPIRED);
```

Comment:
May CAS happened concurrently?
If so, then current thread may return false from this method, however `lastTouched` is already equals EXPIRED.
```
        return currentTimeProvider.now() - last > idleTimeoutMs && 
        (lastTouched.compareAndSet(last, EXPIRED) || lastTouched.get() == EXPIRED);
```
or even just
```
        if (currentTimeProvider.now() - last > idleTimeoutMs)
            lastTouched.set(EXPIRED);
```

Reply:
Good catch! Fixed